File: algoexp/easy/2sum/Solution.go
// # find some elements which are sum of target_sum value
// Time comp.: 0(n*log(n)) Memory comp.: 0(1)

package main

import (
	"fmt"
	"sort"
)

func TwoNumberSum(array []int, target int) []int {
	sort.Ints(array)
	var left int = 0
	var right int = len(array) - 1
	for left < right {
		currentSum := array[left] + array[right]
		if currentSum < target {
			left = left + 1
		} else if currentSum > target {
			right = right - 1
		} else {
			return []int{array[left], array[right]}
		}
	}
	return []int{}
}
func main() {
	result := TwoNumberSum([]int{-7, -5, -3, -1, 0, 1, 3, 5, 7}, -5)
	if len(result) == 0 {
		fmt.Print("where is it?")
	} else {
		fmt.Print(result)
	}
}

File: algoexp/easy/2sum/Solution.py
# find some elements which are sum of target_sum value
# Time comp.: 0(n) Memory comp.: 0(n)


def two_number_sum(array, target_sum):
    elements_d = dict()
    for el in array:
        looking = target_sum - el
        if looking in elements_d:
            return [looking, el]
        else:
            elements_d[el] = el
    return []


if __name__ == '__main__':
    final_arr = two_number_sum(array=[1, 2, -4, 6, 8, -2], target_sum=14)
    if not final_arr:
        print('there are no elements were finded')
    else:
        print("final elements are: " +
              str(final_arr[0]) +
              " and " +
              str(final_arr[1]))
